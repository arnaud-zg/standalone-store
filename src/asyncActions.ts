import { AnyAction, createStore } from 'redux'
import { ActionCreator, isActionOf } from 'typesafe-actions'
import { StandaloneStore } from '.'

interface IResolve<TState> {
  action: AnyAction
  state: TState
}

export const dispatchActionsAndWaitResponse = <TState, TSelectorResponse>({
  actionsDispatch,
  actionCreatorsResolve,
  store,
  selector,
}: {
  store: ReturnType<typeof createStore>
  actionsDispatch: AnyAction[]
  actionCreatorsResolve: ActionCreator[]
  selector: (state: TState) => TSelectorResponse
}) => {
  const standaloneStore = new StandaloneStore<TState>({ store })

  return !actionsDispatch.length
    ? Promise.reject('You should at least give one action.')
    : new Promise((resolve: ({ action, state }: IResolve<TState>) => void) => {
        actionCreatorsResolve.forEach(actionResolve => {
          standaloneStore.subscribe((action, state) => {
            if (isActionOf(actionResolve, action)) {
              resolve({ action, state })
            }
          })
        })
        actionsDispatch.forEach(action => {
          standaloneStore.dispatchAction(action)
        })
      }).then(({ state }: { state: TState }) => {
        standaloneStore.listenersClear()
        return selector(state)
      })
}
