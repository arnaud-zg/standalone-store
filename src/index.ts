export * from './asyncActions'
export * from './standaloneStore'
