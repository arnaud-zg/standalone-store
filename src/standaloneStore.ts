import { AnyAction, StoreCreator } from 'redux'

type Listener<TStoreState> = (
  action: AnyAction | null,
  state: TStoreState
) => void

export class StandaloneStore<TStoreState> {
  private listeners: Array<Listener<TStoreState>>
  private store: ReturnType<StoreCreator>
  private lastAction: AnyAction | null

  constructor({ store }: { store: ReturnType<StoreCreator> }) {
    this.lastAction = null
    this.listeners = []
    this.store = store

    this.store.subscribe(this.storeUpdate)
  }

  dispatchAction = (action: AnyAction) => {
    this.lastAction = action
    this.store.dispatch(action)
  }

  storeUpdate = () => {
    const state = this.store.getState() as TStoreState

    if (this.listeners.length) {
      this.listeners.forEach(listener => {
        listener(this.lastAction, state)
      })
    }
  }

  subscribe(listener: Listener<TStoreState>) {
    this.listeners.push(listener)
  }
}
