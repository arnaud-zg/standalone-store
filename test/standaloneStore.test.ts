import { configureStore, putCredentials, TState } from 'ts-foursquare'
import { getType } from 'typesafe-actions'
import { StandaloneStore } from '../src'

describe('StandaloneStore', () => {
  const store = configureStore({ middlewares: [] })
  it(`should work without any listener`, () => {
    const standaloneStore = new StandaloneStore<TState>({ store })

    standaloneStore.dispatchAction(
      putCredentials({
        clientId: '123',
        clientSecret: '456',
      })
    )
  })

  it(`should make a snapshot of 0 listeners`, () => {
    const standaloneStore = new StandaloneStore<TState>({ store })

    standaloneStore.listenersClear()
    expect(standaloneStore.getListeners().length === 0).toBeTruthy()

    standaloneStore.subscribe(() => {})
    standaloneStore.subscribe(() => {})
    standaloneStore.listenersClear()
    expect(standaloneStore.getListeners().length === 0).toBeTruthy()
  })

  it(`should make a snapshot of registered listeners`, () => {
    const standaloneStore = new StandaloneStore<TState>({ store })
    standaloneStore.subscribe(() => {})

    const listeners = standaloneStore.getListeners()
    expect(listeners).toMatchSnapshot()

    standaloneStore.listenersPop()
    standaloneStore.listenersPop()
  })

  it(`should make a snapshot of store after of action: putCredentials`, done => {
    const standaloneStore = new StandaloneStore<TState>({ store })

    standaloneStore.subscribe((action, state) => {
      expect(state).toMatchSnapshot()

      if (action && action.type === getType(putCredentials)) {
        standaloneStore.listenersPop()
        done()
      }
    })

    standaloneStore.dispatchAction(
      putCredentials({
        clientId: '123',
        clientSecret: '456',
      })
    )
  })
})
