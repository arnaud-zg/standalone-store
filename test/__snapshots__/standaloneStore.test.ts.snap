// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandaloneStore should make a snapshot of store after of action: putCredentials 1`] = `
Object {
  "life": Object {
    "credentials": Object {
      "clientId": "123",
      "clientSecret": "456",
    },
    "status": true,
  },
  "lists": Object {
    "entities": Object {},
  },
  "photos": Object {
    "entities": Object {},
  },
  "request": Object {},
  "status": Object {},
  "venues": Object {
    "categories": Object {},
    "entities": Object {},
    "likesEntities": Object {},
    "miniVenues": Object {},
    "nextVenues": Object {},
    "recommendedPlaces": Object {},
    "similarVenues": Object {},
    "trendingEntities": Object {},
    "venuesListed": Object {},
  },
}
`;
